File: Medium/238_product_array_except_self.py
# Input: nums = [-1,1,0,-3,3]
# Output: [0,0,9,0,0]

class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        pre = [1] * n
        suf = [1] * n
        res = [1] * n

        for i in range(1, n):
            pre[i] = nums[i-1] * pre[i-1]

        for j in range(n-2, -1, -1):
            suf[j] = nums[j+1] * suf[j+1]

        for k in range(n):
            res[k] = pre[k] * suf[k]

        return res

if __name__ == '__main__':
    print(Solution.productExceptSelf(Solution, [1, 2, 3, 4]))

File: Swift-Solution/Medium/238_product_array_except_self.swift
//Example 1:
//
//Input: nums = [1,2,3,4]
//Output: [24,12,8,6]


class Solution {
    func productExceptSelf(_ nums: [Int]) -> [Int] {
        var pre: [Int] = (0...nums.count-1).map { _ in 1 }
        var post: [Int] = (0...nums.count-1).map { _ in 1 }
        var res: [Int] = (0...nums.count-1).map { _ in 1 }

        var tempPre = 1
        var tempPost = 1
        var len = nums.count - 1

        for i in 0...nums.count-1 {
            pre[i] = tempPre
            tempPre = tempPre * nums[i]

            post[len] = tempPost
            tempPost = tempPost * nums[len]
            len -= 1
        }

        for i in (0...nums.count - 1) {
            res[i] = pre[i] * post[i]
        }

        return res
    }
}